yellow-paper/docs/circuits/private-kernel-tail.md

```diff
@@ -0,0 +1,224 @@
+
+- [Initial private kernel proof](./private-kernel-initial.md).
+- [Inner private kernel proof](./private-kernel-inner.md).
+- [Reset private kernel proof](./private-kernel-reset.md).
+
+An inner iteration may be omitted when there's only a single private function call for the transaction. And a reset iteration can be skipped if there are no read requests and transient nullifiers in the public inputs from the last initial or inner iteration.
+
+#### Ensuring the previous iteration is the last.
+
+The following must be empty to ensure all the private function calls are processed:
+
+- Private call requests.
+
+The following must be empty to ensure a comprehensive final reset:
+
+- The nullified note hash associated with each nullifier.
+- Read requests.
+
+> A [reset iteration](./private-kernel-reset.md) should ideally precede this step. Although it doesn't have to be executed immediately before the tail circuit, as long as it effectively clears the specified values.
+
+### Processing Final Outputs
+
+#### Siloing values.
+
+1. This circuit must silo the following with each item's contract address:
+
+   - Note hash contexts.
+   - Nullifier contexts.
+
+   The siloed value is computed as: `hash(contract_address, value)`.
+
+   Siloing with a contract address ensures that data produced by a contract is accurately attributed to the correct contract and cannot be misconstrued as data created in a different contract.
+
+2. The circuit then applies nonces to the note hashes:
+
+   - The nonce for a note hash is computed as: `hash(first_nullifier, index)`, where:
+     - `first_nullifier` is the [hash of the transaction request](./private-kernel-initial.md#ensuring-transaction-uniqueness).
+     - `index` is the position of the note hash in the note hashes array in the public inputs.
+
+   Siloing with a nonce guarantees that each final note hash is a unique value in the note hash tree.
+
+3. Additionally, this circuit generates the final hashes for L2-L1 messages, calculated as:
+
+   `hash(contract_address, version_id, portal_contract_address, chain_id, message)`
+
+   Where _version_id_ and _portal_contract_address_ equal the values defined in the constant data.
+
+#### Verifying ordered arrays.
+
+The initial and inner kernel iterations may produce values in an unordered state due to the serial nature of the kernel, contrasting with the stack-based nature of code execution.
+
+This circuit ensures the correct ordering of the following arrays (_ordered_array_) in public inputs:
+
+- Note hashes.
+- Nullifiers.
+- L2-to-l1 messages.
+- Public call requests.
+- New contracts (if public call request is empty).
+- New contract contexts (if public call request is not empty).
+
+The corresponding _unordered_arrays_ for the above are sourced either from the transient accumulated data of the previous iteration or from the [siloed results](#siloing-values).
+
+A hints array is provided through private inputs for every _unordered_array_.
```

Comment:
It would be helpful to define (somewhere in the yellow paper, maybe no necessarily here) what we explicitly mean by "hint"

i.e. information provided by the Prover via private inputs. The Prover may choose to provide inaccurate hints, but doing so should produce unsatisfiable constraints and a failing proof. (does this align with how "hint" is being used here?)